The general algorithm

crisely09 opened the pull request:
There is probably a better way to select the conditions...

crisely09 committed:
Add simple/generic algorithm for line search

1. Add naive line search algorithm
2. Add test

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+        Direction vector for next step.
+    alpha: float
+        The current step length estimate
+    conditions: str or list of str
```
{str, list of str}

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+    conditions: str or list of str
+        Conditions to be satisfied by the new step length.
+        Valid options are: 'soft-wolfe', 'strong-wolfe' and 'armijo'
+    const1: float
```
{float, 1e-4}

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+        Valid options are: 'soft-wolfe', 'strong-wolfe' and 'armijo'
+    const1: float
+        Constant c_1 for Wolfe/Armijo conditions
+    const2: float
```
{float, 1e-2}

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+        raise ValueError("Constants c_1 and c_2  must meet the condition:"
+                         " 0 < c_1 < c_2 < 1.0")
+    # Check duplicates if there are any
+    if len(conditions) > 1:
```
if len(conditions) == len(set(conditions))

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+    # Check valid conditions
+    for i, condition in enumerate(conditions):
+        if condition not in ['soft-wolfe', 'strong-wolfe', 'armijo']:
+            raise ValueError("Condition %d+1 is not implemented." % i)
```
raise ValueError('Condition, {}, is not supported'.format(condition))
remove enumerate

kimt33 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+                    return alpha_new
+            elif condition == 'armijo':
+                if armijo(func, grad, var, direction, alpha_new):
+                    return alpha_new
```
The algorithm doesn't quite do what it says it'll do. It doesn't make sure that all of the conditions are satisfied. It just returns the first alpha that satisfies one of the conditions.

kimt33 commented:
I've made the requested changes. It doesn't seem to converge though. I think it has to do with the alpha update (because it gets stuck at 0.5).

crisely09 commented on flik/linesearch/linesearch.py:
```diff
@@ -0,0 +1,76 @@
+    while True:
+        alpha_new = cubic_interpolation(var, func, grad, direction, alpha, alpha_new)
+        for condition in conditions:
+            if satisfies_conditions(alpha_new):
```
This is definitely a nicer way to do it.

kimt33 committed:
Fix bug in conditions

1. armijo condition had a bug
2. hard coded fix was removed from linesearch algorithm

kimt33 commented:
I'm merging this, even though the line search module is not complete. This way, we can have all the code available to us when we refactor the line search module.
